File: 128.longest-consecutive-sequence.java

// @leet imports start
import java.util.*;
import java.math.*;
// @leet imports end

// @leet start
class Solution {
    public int longestConsecutive(int[] nums) {
        Map<Integer, Integer> m = new HashMap<>();
        int ans = 0;
        for (int i : nums) {
            if (!m.containsKey(i)) {
                // connect m[i] to m[i+1] and m[i-1]
                int less = m.getOrDefault(i - 1, 0);
                int more = m.getOrDefault(i + 1, 0);
                int curr = less + more + 1;
                m.put(i, curr);
                // update the boundary values
                if (less > 0) {
                    m.put(i - less, curr);
                }
                if (more > 0) {
                    m.put(i + more, curr);
                }
                ans = Math.max(ans, curr);
            }
        }
        return ans;

    }
}
// @leet end

File: 128.longest-consecutive-sequence.py
from collections import Counter
from typing import List


# @leet start
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        ans = 0
        c = set(nums)
        starts = set()
        for e in c:
            if e - 1 not in c:
                starts.add(e)

        for i in starts:
            chain = 0
            while i in c:
                i += 1
                chain += 1
            ans = max(chain, ans)

        return ans


# @leet end

